src/lib/relink_private.js

```diff
@@ -37,9 +37,14 @@ module.exports = function relinkPrivateKeys(toContainer, fromContainer) {
         }
         else if(isArray(fromVal) && isArray(toVal) && isPlainObject(fromVal[0])) {
 
+            // filter out data_array items that can contain user objects
+            // most of the time the toVal === fromVal check will catch these early
+            // but if the user makes new ones we also don't want to recurse in.
+            if(k === 'customdata' || k === 'ids') continue;
```

Comment:
For reference, you should probably right down somewhere that values in the `ids` arrays should be strings.

But yeah, :+1: for being safe.

Reply:
> For reference, we should probably right down somewhere that values in the `ids` arrays should be strings.

I can put that in its attribute `description`, but is it important that these are strings and not numbers? Do they get converted to actual strings during use?

Reply:
> Do they get converted to actual strings during use?

yep: `cd[i].id = String(trace.ids[i]);`

Reply:
> I can put that in its attribute description

That would be great!

Reply:
comment added in https://github.com/plotly/plotly.js/pull/2375/commits/62e54b70c87f8438d2ebd64fb8db90dc8b926785